File: frontend/src/models/Settlement.ts
import {TileIdentifier} from "./tile";
import {CountryIdentifier} from "./country";
import {Color} from "./color";

export interface Settlement {
	identifier: SettlementIdentifier,
	country: CountryIdentifier
	tile: TileIdentifier,
}

export interface SettlementIdentifier {
	id: string,
	name: string
	color: Color,
}

File: backend/strategy-game-common-data/src/main/kotlin/io/github/smiley4/strategygame/backend/commondata/Settlement.kt
package io.github.smiley4.strategygame.backend.commondata


class Settlement(
    val id: Id,
    val country: Country.Id,
    val tile: TileRef,
    val attributes: Attributes,
    val infrastructure: Infrastructure,
    val population: Population,
    var resourceLedger: ResourceLedger
) {

    @JvmInline
    value class Id(val value: String) {
        companion object
    }

    class Attributes(
        val name: String,
        val color: RGBColor,
        val viewDistance: Int,
    )

    class Infrastructure(
        val productionQueue: MutableList<ProductionQueueEntry>,
        val buildings: MutableList<Building>,
    )

    class Population(
        var size: Int,
        var growthProgress: Float,
        var growthAmount: Float,
        val growthDetails: MutableMap<String, Float> = mutableMapOf(),
    )

}
